internal/analytics/deferred.go

```diff
@@ -0,0 +1,62 @@
+package analytics
+
+import (
+	"encoding/json"
+
+	"github.com/spf13/viper"
+
+	"github.com/ActiveState/cli/internal/logging"
+)
+
+var Defer bool
+
+type deferredData struct {
+	Category   string
+	Action     string
+	Label      string
+	Dimensions map[string]string
+}
+
+const deferredCfgKey = "deferrer_analytics"
+
+func deferEvent(category, action, label string, dimensions map[string]string) error {
+	logging.Debug("Deferring: %s, %s, %s", category, action, label)
+	deferred := loadDeferred()
+	deferred = append(deferred, deferredData{category, action, label, dimensions})
+	saveDeferred(deferred)
+	viper.WriteConfig() // the global viper instance is bugged, need to work around it for now -- https://www.pivotaltracker.com/story/show/175624789
+	return nil
+}
+
+func sendDeferred(sender func(string, string, string, map[string]string) error) error {
+	deferred := loadDeferred()
+	for n, event := range deferred {
+		if err := sender(event.Category, event.Action, event.Label, event.Dimensions); err != nil {
+			return err
+		}
+		saveDeferred(deferred[n+1:])
+	}
```

Comment:
The control flow this creates is hard for me to grok right now, but I think this is going to result in some potentially intense churning of the config and possibly duplicated/bad data. The deferred data is loaded and iterated over, then each value is appended back to memory and disk, then a different set of data is stored to memory, then the result of all those iterations is stored to disk. I think it may be "fine", but an error at line 35 here will likely make the retained data incorrect.

Reply:
`saveDeferred` doesn't write to disk. It just saves the config in memory. Same thing for `loadDeferred` if it was already previously loaded.

Reply:
`sender` is intended to be deferEvent at times, so I believe that my assessment is correct and an error will result in duplicated/bad data.